[Diagnostic] Fix diagnostic when checking conformance for IUO of gene…

srinikhil-07 opened the pull request:
…ric parameter

Prints TypeRepr in diagnostic if possible to throw
accurate diagnostic for IUO conformance checking
Fixes [rdar://problem/64953106]. Fixes SR-13119

<!-- What's in this pull request? -->
The diagnostic emitted when checking conformance for IUO of generic parameter will
print TypeRepr if present to have accurate type information. This prevents 
confusion in the diagnostics emitted.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Fixes [rdar://problem/64953106]. Fixes SR-13119(https://bugs.swift.org/browse/SR-13119).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

srinikhil-07 committed:
[Diagnostic] Fix diagnostic when checking conformance for IUO of generic parameter

Prints TypeRepr in diagnostic if possible to throw
accurate diagnostic for IUO conformance checking
Fixes [rdar://problem/64953106]. Fixes SR-13119

LucianoPAlmeida commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
 
       case DoesNotConform:
+        // for accurate diagnostic, we show TypeRepr if present for IUO
+        TypeLoc typeLoc = TypeLoc();
```
I think this can be simplified as `TypeLoc typeLoc;`

LucianoPAlmeida commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
       case DoesNotConform:
+        // for accurate diagnostic, we show TypeRepr if present for IUO
+        TypeLoc typeLoc = TypeLoc();
+        if(varDecl->isImplicitlyUnwrappedOptional()){
```
```suggestion
        if(varDecl->isImplicitlyUnwrappedOptional()) {
```

CodaFi replied:
In general, please run `clang-format` on this buffer.

theblixguy commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
+                NULL,
+                varDecl->getType(),
+            };
+        }
```
Given we're doing this in three places in this function, I think it would be a good idea to extract and reuse:

```c++
static TypeLoc getTypeLocForDiagnostics(VarDecl *VD) {
  auto varTy = VD->getType();
  if (VD->isImplicitlyUnwrappedOptional()) {
    return TypeLoc(VD->getTypeReprOrParentPatternTypeRepr(), varTy);
  }
  return TypeLoc::withoutLoc(varTy);
}
```

CodaFi replied:
I don't think this needs to be a utility. Its ideal form is four lines

```
TypeLoc typeLoc = {
  varDecl->getTypeReprOrParentPatternTypeRepr(),
 varDecl->getType(),
};
```

CodaFi commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
+            };
+        } else {
+            typeLoc = {
+                NULL,
```
You don't need to conditionalize this. `getTypeReprOrParentPatternTypeRepr` will always get what the user wrote, if they wrote something at all.

Nikhil0487 replied:
@CodaFi 

Yes, but this is breaking nested class/struct types. Please see the test regression I have commented in the https://bugs.swift.org/browse/SR-13119. @varungandhi-apple suggested a change which I inferred to be the above.

CodaFi replied:
Name lookup will see the innermost name in that scope and it should just work out. As long as those fixits compile, this is not a regression and we should update those tests. I very much prefer we preserve the way the user spelled types than fully qualify them all the time.

Nikhil0487 replied:
Sure, okay. I get it

varungandhi-apple replied:
> I very much prefer we preserve the way the user spelled types than fully qualify them all the time.

I didn't realize there was a disagreement here. Should we write this down somewhere so that newer diagnostics use this style? I thought this was a regression because the original code was printing the full type.

CodaFi replied:
That's a good point. I recall we have a document that tries to explain good diagnostic hygiene. And this is certainly one of those places where you have to be really careful. Because the user-provided type may not always work if you're pulling something from a different DeclContext, and the fully-qualified type may not even work when you have ambiguities caused by e.g. the module-and-type-sharing-names bug.

CodaFi reviewed (changes requested):
A couple of nits, then this is good to merge.

srinikhil-07 committed:
Nested name lookup tests update

Name lookup will see the innermost name anyway and
preferred over fully qualified name. Hence the test
cases are also updated.

LucianoPAlmeida commented on test/decl/protocol/special/coding/struct_codable_failure_diagnostics.swift:
```diff
@@ -9,8 +9,8 @@ struct S1 : Codable {
-  // expected-note@-1 {{cannot automatically synthesize 'Decodable' because 'S1.Nested' does not conform to 'Decodable'}}
-  // expected-note@-2 {{cannot automatically synthesize 'Encodable' because 'S1.Nested' does not conform to 'Encodable'}}
+  // expected-note@-1 {{cannot automatically synthesize 'Decodable' because 'Nested' does not conform to 'Decodable'}}
+  // expected-note@-2 {{cannot automatically synthesize 'Encodable' because 'sNested' does not conform to 'Encodable'}}
```
Typo 'sNested' -> 'Nested'?

srinikhil-07 committed:
Replaced a letter in test case that inadvertently got added

varungandhi-apple commented:
Did you try using `Type::getString` with `PrintOptionalAsImplicitlyUnwrapped` set to `true`? Does that avoid the regression for the nested type diagnostics?

theblixguy commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -173,8 +173,13 @@ static bool validateCodingKeysEnum(DerivedConformance &derived,
         break;
 
       case DoesNotConform:
+        // for accurate diagnostic, we show TypeRepr if present
```
It might be a good idea to rephrase this to provide better context. It could be something along the lines of:

```suggestion
        // We use a TypeLoc here so diagnostics can show the type
        // as written by the user in source if possible. This is useful
        // when the user has written an IUO type for example, since
        // diagnostics would show the type as 'T?' instead of 'T!' if
        // we use a Type instead.
```

Nikhil0487 replied:
Thanks @theblixguy, will update at both the places.

srinikhil-07 committed:
Code format, corrections and better comments

This commit includes better comments for easy
understanding, formatted the bug fix code with
clang-format and fixes wrong variables inadvertently
introduced.

CodaFi reviewed (approved):
Very nice! Thanks ⛵

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

LucianoPAlmeida commented:
@Nikhil0487  I think we could update those tests that are failing because although they are misleading because `Int` does conform with codable. They probably will be removed in #32371 anyways ... I can rebase those changes there no problem :)

srinikhil-07 committed:
[Test] Update type in struct codable test

This commit changes diagnostic type from error type
to Int. Although this diagnostic updated is incorrect, this will
be resolved when 32371 gets pulled.

LucianoPAlmeida commented:
@swift-ci Please smoke test

LucianoPAlmeida commented:
@swift-ci Please smoke test macOS platform
